evennia/web/website/views.py

```diff
@@ -689,6 +690,168 @@ def form_valid(self, form):
             messages.error(self.request, "Your character could not be created.")
             return self.form_invalid(form)
             
+#
+# Channel views
+#
+
+class ChannelMixin(object):
+    """
+    This is a "mixin", a modifier of sorts.
+    
+    Any view class with this in its inheritance list will be modified to work
+    with HelpEntry objects instead of generic Objects or otherwise.
+
+    """
+    # -- Django constructs --
+    model = class_from_module(settings.BASE_CHANNEL_TYPECLASS)
+    
+    # -- Evennia constructs --
+    page_title = 'Channels'
+
+    # What lock type to check for the requesting user, authenticated or not.
+    # https://github.com/evennia/evennia/wiki/Locks#valid-access_types
+    access_type = 'listen'
+    
+    def get_queryset(self):
+        """
+        Django hook; here we want to return a list of only those Channels
+        and other documentation that the current user is allowed to see.
+        
+        Returns:
+            queryset (QuerySet): List of Channels available to the user.
+            
+        """
+        account = self.request.user
+        
+        # Get list of all Channels
+        channels = self.model.objects.all().iterator()
+        
+        # Now figure out which ones the current user is allowed to see
+        bucket = [channel.id for channel in channels if channel.access(account, 'listen')]
+        
+        # Re-query and set a sorted list
+        filtered = self.model.objects.filter(
+            id__in=bucket
+        ).order_by(
+            Lower('db_key')
+        )
+        
+        return filtered
+        
+
+class ChannelListView(ChannelMixin, ListView):
+    """
+    Returns a list of channels that can be viewed by a user, authenticated
+    or not.
+    
+    """
+    # -- Django constructs --
+    paginate_by = 100
+    template_name = 'website/channel_list.html'
+    
+    # -- Evennia constructs --
+    page_title = "Channel Index"
+    
+    max_popular = 10
+    
+    def get_context_data(self, **kwargs):
+        """
+        Django hook; we override it to calculate the most popular channels.
+        
+        Returns:
+            context (dict): Django context object
+        
+        """
+        context = super(ChannelListView, self).get_context_data(**kwargs)
+        
+        # Calculate which channels are most popular
+        context['most_popular'] = sorted(
+            list(self.get_queryset()), 
+            key=lambda channel: len(channel.subscriptions.all()), 
```

Comment:
Under normal circumstances I'd agree, but *.subscriptions.all() doesn't return a queryset-- it returns a dict (?!)

```
>>> Channel.create('demo')
(<Channel: demo>, [])
>>> c = Channel.objects.first()
>>> c.subscriptions.all().count()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'dict' object has no attribute 'count'
```

Reply:
Hm, so it seems. The reason would be because the `subscriptions` handler actually deals with two model-types at the same time (Accounts and Objects) so its abstracting away those underlying fields. Since the lookup is heavily cached I suppose using `len()` is indeed to go without starting to dive into the underlying fields. Ok then!